From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samuel Attard <marshallofsound@electronjs.org>
Date: Wed, 26 Jul 2023 17:03:15 -0700
Subject: fix: do not resolve electron entrypoints

This wastes fs cycles and can result in strange behavior if this path actually exists on disk

diff --git a/lib/internal/modules/esm/load.js b/lib/internal/modules/esm/load.js
index 8406e8ca319533d6186c5d457e951b7a845b6b72..caa34187e17d9a0addcc8c94e46823ea4e05c590 100644
--- a/lib/internal/modules/esm/load.js
+++ b/lib/internal/modules/esm/load.js
@@ -131,7 +131,7 @@ async function defaultLoad(url, context = kEmptyObject) {
     format ??= 'builtin';
   } else {
     let contextToPass = context;
-    if (source == null) {
+    if (format !== 'electron' && source == null) {
       ({ responseURL, source } = await getSource(urlInstance, context));
       contextToPass = { __proto__: context, source };
     }
@@ -139,7 +139,7 @@ async function defaultLoad(url, context = kEmptyObject) {
     // Now that we have the source for the module, run `defaultGetFormat` again in case we detect ESM syntax.
     format ??= await defaultGetFormat(urlInstance, contextToPass);
 
-    if ((format === 'commonjs' || format === 'electron') && contextToPass !== context) {
+    if (format === 'electron' || format === 'commonjs' && contextToPass !== context) {
       // For backward compatibility reasons, we need to discard the source in
       // order for the CJS loader to re-fetch it.
       source = null;
diff --git a/lib/internal/modules/esm/translators.js b/lib/internal/modules/esm/translators.js
index dbbe37474c70addfbf8cf805c29732dcd38c74f6..96ffb2b0b051163d8a777ed3d1b5bc2c83fc37c5 100644
--- a/lib/internal/modules/esm/translators.js
+++ b/lib/internal/modules/esm/translators.js
@@ -376,6 +376,9 @@ function cjsPreparseModuleExports(filename, source) {
     if (cached) {
       return { module, exportNames: cached.exportNames };
     }
+    if (filename === 'electron') {
+      return { module, exportNames: new SafeSet(['default', ...Object.keys(module.exports)]) };
+    }
   }
   const loaded = Boolean(module);
   if (!loaded) {
diff --git a/lib/internal/modules/run_main.js b/lib/internal/modules/run_main.js
index 1f03c313121db054ff824d07e57c57c749005497..2c8aa6461497f39062ec350ef8a063b9b0ac1edf 100644
--- a/lib/internal/modules/run_main.js
+++ b/lib/internal/modules/run_main.js
@@ -2,6 +2,7 @@
 
 const {
   StringPrototypeEndsWith,
+  StringPrototypeStartsWith,
 } = primordials;
 
 const { containsModuleSyntax } = internalBinding('contextify');
@@ -13,6 +14,13 @@ const path = require('path');
  * @param {string} main - Entry point path
  */
 function resolveMainPath(main) {
+  // For built-in modules used as the main entry point we _never_
+  // want to waste cycles resolving them to file paths on disk
+  // that actually might exist
+  if (typeof main === 'string' && StringPrototypeStartsWith(main, 'electron/js2c')) {
+    return main;
+  }
+
   const defaultType = getOptionValue('--experimental-default-type');
   /** @type {string} */
   let mainPath;
@@ -50,6 +58,13 @@ function resolveMainPath(main) {
  * @param {string} mainPath - Absolute path to the main entry point
  */
 function shouldUseESMLoader(mainPath) {
+  // For built-in modules used as the main entry point we _never_
+  // want to waste cycles resolving them to file paths on disk
+  // that actually might exist
+  if (typeof mainPath === 'string' && StringPrototypeStartsWith(mainPath, 'electron/js2c')) {
+    return false;
+  }
+
   if (getOptionValue('--experimental-default-type') === 'module') { return true; }
 
   /**
